File: ruby/lib/runapi/nano_banana/resources/edit_image.rb
# frozen_string_literal: true

module RunApi
  module NanoBanana
    module Resources
      # NanoBanana image editing resource.
      # Edit existing images using text prompts and reference images.
      class EditImage
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/nano_banana/edit_image"

        RESPONSE_CLASS = Types::EditImageResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedEditImageResponse

        def initialize(http)
          @http = http
        end

        # Edit an image and wait until complete.
        #
        # @param params [Hash] edit parameters
        # @return [RunApi::NanoBanana::Types::CompletedEditImageResponse] completed edit with image results
        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        # Create an image edit task.
        #
        # @param params [Hash] edit parameters
        # @return [RunApi::NanoBanana::Types::EditImageResponse] task creation result with id
        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        # Get edit status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::NanoBanana::Types::EditImageResponse] current edit status
        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["edit-image"], params)

          raise Core::ValidationError, "prompt is required" unless param(params, :prompt)
          raise Core::ValidationError, "source_image_urls is required" unless param(params, :source_image_urls)
        end
      end
    end
  end
end

File: python/src/runapi/nano_banana/resources/edit_image.py
"""Nano Banana image editing resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..contract_gen import CONTRACT
from ..types import (
    CompletedEditImageResponse,
    EditImageResponse,
)


class EditImage(Resource):
    """Edit existing images using text prompts and reference images."""

    ENDPOINT = "/api/v1/nano_banana/edit_image"

    RESPONSE_CLASS = EditImageResponse
    COMPLETED_RESPONSE_CLASS = CompletedEditImageResponse

    def run(self, **params: Any) -> Any:
        """Create an edit-image task and poll until it completes.

        Args:
            **params: Edit-image parameters (model, prompt, source_image_urls, ...).

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create an edit-image task and return immediately with an ``id``.

        Args:
            **params: Edit-image parameters (model, prompt, source_image_urls, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of an edit-image task.

        Args:
            id: The task id.

        Returns:
            The current status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["edit-image"], params)

        if not params.get("prompt"):
            raise ValidationError("prompt is required")
        if not params.get("source_image_urls"):
            raise ValidationError("source_image_urls is required")
